src/meta/client/MetaClient.cpp

```diff
@@ -107,6 +110,35 @@ void MetaClient::stop() {
     isRunning_ = false;
 }
 
+void MetaClient::dnsQUeryThreadFunc() {
```

Comment:
I am a little confused about this function. 
`addrs_` contains all meta servers addresses and all of them are represented by hostname currently,  right?
If so,  why we resolve them again?  The connection will be rebuild in ThriftClientManager

Reply:
The IP address behind the domain name may change, and thrift does not automatically resolve the corresponding domain name.

@knightXun

Reply:
> thrift does not automatically resolve the corresponding domain name.

I notice the ctor about folly::SocketAddress looks like 
```
 SocketAddress(const char* host, uint16_t port, bool allowNameLookup = false) 
```
So the third params means we could resolve the dns?

Reply:
Yes, this parameter is to decide whether to perform dns lookup when constructing folly :: SocketAddress.

Reply:
So why not resolve the dns when rebuilding connection.

Reply:
This DNS lookup is performed only when the constructor is called, that is, after the object is constructed, the DNS lookup is no longer implicitly performed.

Reply:
There are many ways to solve the problem.  For example, you could reset the original 'SocketAddress' instance or construct a new one. 

We don't need the loop here.

Reply:
> There are many ways to solve the problem. For example, you could reset the original 'SocketAddress' instance or construct a new one.
> 
> We don't need the loop here.

First of all, dns query is blocking, so we must minimize the impact of blocking, so we need to do it in a separate thread.
Secondly, we don't know when the dns lookup is needed, so here we need to loop to keep looking up.

Reply:
> First of all, dns query is blocking, so we must minimize the impact of blocking, so we need to do it in a separate thread.

Only build a new connection needs the dns LookUp. I think it is ok for the blocking. 

> Secondly, we don't know when the dns lookup is needed, so here we need to loop to keep looking up.

Actually, we know it.  It will be used on cloud, that means only the old host dead and replace the old host with a new one.  So when the old connection broken down,  we should retry to resolve the dns. 

The loop for dns resolve is not a good way.  The main reasons are:
1. The loop here is useless in most time because the hosts IPs are not always changed. It costs too much.
2. How to deal with the storaged hosts IPs with your approach?


WDTY?

Reply:
> > First of all, dns query is blocking, so we must minimize the impact of blocking, so we need to do it in a separate thread.
> 
> Only build a new connection needs the dns LookUp. I think it is ok for the blocking.
> 
> > Secondly, we don't know when the dns lookup is needed, so here we need to loop to keep looking up.
> 
> Actually, we know it. It will be used on cloud, that means only the old host dead and replace the old host with a new one. So when the old connection broken down, we should retry to resolve the dns.
> 
> The loop for dns resolve is not a good way. The main reasons are:
> 
> 1. The loop here is useless in most time because the hosts IPs are not always changed. It costs too much.
1. If only a certain IP address has a temporary network failure, then we do not necessarily need to re-block to do a DNS query.
2. In most cases, the DNS query is useless, but the current implementation will not affect any other operation of the system, because the corresponding addrs_ structure will be updated only when the IP address changes.
> 2. How to deal with the storaged hosts IPs with your approach?
This issue has not been resolved for the time being, I will fix it next
> 
> WDTY?

Reply:
If you really worry about the blocking problem.  I suggest you put the loop inside ThriftClientManager,  because it has hold all hosts used in current service. And it could handle the storaged IP changes too.

Reply:
I think ThriftClientManager is better to handle IP addresses only.

1. currently we only need to dynamically handle the domain name of the meta server, because storage sends its own IP to the meta server every time the storage is started, which means we do not support a storage server that does not restart and change its IP address.
2 Since clientMap_ is ThreadLocal, if you want to update the IP of a domain name in ThriftClientManager, it means we need to update clientMap_ in all threads within one thread, which I don't think is necessary.